src/Query.php

```diff
@@ -345,6 +342,84 @@ protected function _render_from()
 
     /// }}}
 
+    // {{{ with()
+
+    /**
+     * Specify WITH query to be used.
+     *
+     * @param Query  $cursor    Specifies cursor query or array [alias=>query] for adding multiple
+     * @param string $alias     Specify alias for this cursor
+     * @param array  $fields    Optional array of field names used in cursor
+     * @param bool   $recursive Is it recursive?
+     *
+     * @return $this
+     */
+    public function with(self $cursor, string $alias, ?array $fields = null, bool $recursive = false)
+    {
+        // save cursor in args
+        $this->_set_args('with', $alias, [
+            'cursor'    => $cursor,
+            'fields'    => $fields,
+            'recursive' => $recursive,
+        ]);
+
+        return $this;
+    }
+
+    /**
+     * Recursive WITH query.
+     *
+     * @param Query|array $cursor Specifies cursor query or array [alias=>query] for adding multiple
+     * @param string      $alias  Specify alias for this cursor
+     * @param array       $fields Optional array of field names used in cursor
+     *
+     * @return $this
+     */
+    public function withRecursive(self $cursor, string $alias, ?array $fields = null)
+    {
+        return $this->with($cursor, $alias, $fields, true);
+    }
+
+    /**
+     * Renders part of the template: [with]
+     * Do not call directly.
+     *
+     * @return string Parsed template chunk
+     */
+    protected function _render_with()
+    {
+        // will be joined for output
+        $ret = [];
+
+        if (empty($this->args['with'])) {
+            return '';
+        }
+
+        // process each defined cursor
+        $isRecursive = false;
+        foreach ($this->args['with'] as $alias => ['cursor'=>$cursor, 'fields'=>$fields, 'recursive'=>$recursive]) {
+            // cursor alias cannot be expression, so simply escape it
+            $s = $this->_escape($alias).' ';
+
+            // set cursor fields
+            if ($fields) {
+                $s .= '('.implode(',', array_map([$this, '_escape'], $fields)).') ';
+            }
+
+            // will parameterize the value and escape if necessary
+            $s .= 'as '.$this->_consume($cursor, 'soft-escape');
```

Comment:
`AS` in UC would be nicer

Reply:
we don't use UC in SQL. I find this "standard" old and silly. I understand not all computers have lower case characters, but we don't see many of them anymore :D

Reply:
I like UC too, but made it lowercase for consistency. See - all templates currently use lowercase. If we update it to UC, then should do that in all templates not just this one.